File: starpu/parse_dag/parse_deps.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <algorithm>
#include <set>
#include <map>
#include <queue>

#define DEBUG 0

class TaskInfo {
public:
    TaskInfo(const std::string& dag_file) : dag_file(dag_file) {
        summary();
    }

    void summary() {
        parse_dag();
        topological_sort();
        reindex_task();
    }
    using TaskDep = std::vector<std::pair<int, int>>;
    using LayerDep = std::vector<TaskDep>;
    using GraphDep = std::vector<LayerDep>;

    GraphDep get_dep() {
        GraphDep result;
        for (const auto& layer : layer_topo_order) {
            LayerDep layer_deps;
            for (const auto& task : layer) {
                auto dep = get_dep(task);
                layer_deps.push_back(dep);
            }
            result.push_back(layer_deps);
        }
        return result;
    }

    int get_timestamp() const {
        return layer_topo_order.size();
    }

    int get_width_of_timestamp(int t) const {
        return layer_topo_order[t].size();
    }

    const std::vector<std::string>& get_task_of_timestamp(int t) const {
        return layer_topo_order[t];
    }

    friend std::ostream& operator<<(std::ostream& os, const TaskInfo& task_info);

private:
    std::string dag_file;
    std::vector<std::string> all_task;
    std::set<std::string> input_task;
    std::set<std::string> output_task;
    std::map<std::string, std::vector<std::string>> task2output;
    std::map<std::string, std::string> tag2task;
    std::map<std::string, std::string> task2tag;
    std::set<std::string> has_input;
    std::map<std::string, std::vector<std::string>> task2input;
    std::map<std::string, std::vector<std::string>> task2input_copy;
    std::vector<std::vector<std::string>> layer_topo_order;
    std::map<std::string, std::pair<int, int>> task2index;

    void parse_dag() {
        std::ifstream file(dag_file);
        std::string line;
        std::vector<std::string> lines;

        while (std::getline(file, line)) {
            lines.push_back(line);
        }

        bool parse_tag = true;
        int split_line = -1;

        for (int i = 0; i < lines.size(); ++i) {
            line = lines[i];
            line = trim(line);
            std::cout << line << std::endl;

            if (line.empty()) {
                split_line = i;
                break;
            }
        }

        std::vector<std::string> tag_info(lines.begin(), lines.begin() + split_line);
        std::vector<std::string> dep_info(lines.begin() + split_line + 1, lines.end());

        for (const auto& tag_line : tag_info) {
            std::string tag, tasks_str;
            // split by ":"
            int pos = tag_line.find(":");
            tag = tag_line.substr(0, pos);
            tasks_str = tag_line.substr(pos + 1);
            tag = trim(tag);
            tasks_str = trim(tasks_str, " []");
            std::cout << tag << " " << tasks_str << std::endl;

            std::istringstream tasks_ss(tasks_str);
            std::string task;
            while (std::getline(tasks_ss, task, ',')) {
                task = trim(task);
                all_task.push_back(task);
                tag2task[tag] = task;
                task2tag[task] = tag;
            }
        }

        for (const auto& dep_line : dep_info) {
            std::string task, outputs_str;
            // split by ":"
            int pos = dep_line.find(":");
            task = dep_line.substr(0, pos);
            outputs_str = dep_line.substr(pos + 1);

            task = trim(task);
            outputs_str = trim(outputs_str, "(),");

            std::istringstream outputs_ss(outputs_str);
            std::string output;
            while (std::getline(outputs_ss, output, ',')) {
                output = trim(output);
                task2output[task].push_back(output);
                has_input.insert(output);
                task2input[output].push_back(task);
            }

            if (task2output[task].empty()) {
                output_task.insert(task);
            }
        }

        std::set<std::string> all_task_set(all_task.begin(), all_task.end());

        std::set_difference(all_task_set.begin(), all_task_set.end(), has_input.begin(), has_input.end(),
                            std::inserter(input_task, input_task.begin()));

        task2input_copy = task2input;
    }

    // std::set<T> set_difference(const std::set<T>& set1, const std::set<T>& set2) {
    //     std::set<T> result;
    //     std::set_difference(set1.begin(), set1.end(), set2.begin(), set2.end(), std::inserter(result, result.begin()));
    //     return result;
    // }

    void topological_sort() {
        std::queue<std::string> task_queue;
        std::cout << "input_task: " << std::endl;
        for (const auto& task : input_task) {
            task_queue.push(task);
        }

        while (!task_queue.empty()) {
            int n = task_queue.size();
            std::vector<std::string> layer;

            for (int i = 0; i < n; ++i) {
                std::string task = task_queue.front();
                task_queue.pop();
                layer.push_back(task);

                const auto& outputs = task2output[task];
                for (const auto& output : outputs) {
                    // remove will move task into the end of vector, and return the iterator of the end
                    // then call erase to remove the task from the vector
                    task2input[output].erase(std::remove(task2input[output].begin(), task2input[output].end(), task),
                                             task2input[output].end());

                    if (task2input[output].empty()) {
                        task_queue.push(output);
                    }
                }
            }

            layer_topo_order.push_back(layer);
        }
    }

    TaskDep get_dep(const std::string& task) const {
        if (input_task.count(task) > 0) {
            return {std::make_pair(-1, -1)};
        }

        TaskDep result;
        const auto& inputs = task2input_copy.at(task);
        for (const auto& input : inputs) {
            const auto& index = task2index.at(input);
            result.push_back(index);
        }

        return result;
    }

    void reindex_task() {
        for (int t = 0; t < layer_topo_order.size(); ++t) {
            const auto& layer = layer_topo_order[t];
            for (int w = 0; w < layer.size(); ++w) {
                const std::string& task = layer[w];
                task2index[task] = std::make_pair(t, w);
            }
        }
    }

    std::string trim(const std::string& str, const std::string& delimiters = " \f\n\r\t\v") const {
        size_t first = str.find_first_not_of(delimiters);
        size_t last = str.find_last_not_of(delimiters);
        return (first == std::string::npos || last == std::string::npos) ? "" : str.substr(first, last - first + 1);
    }
};

std::ostream& operator<<(std::ostream& os, const TaskInfo& task_info) {
    os << "dag_file: " << task_info.dag_file << "\n"
       << "all_task: ";
    for (const auto& task : task_info.all_task) {
        os << task << " ";
    }
    os << "\n"
       << "input_task: ";
    for (const auto& task : task_info.input_task) {
        os << task << " ";
    }
    os << "\n"
       << "output_task: ";
    for (const auto& task : task_info.output_task) {
        os << task << " ";
    }
    os << "\n"
       << "task2output: {";
    for (const auto& entry : task_info.task2output) {
        os << entry.first << ": [";
        for (const auto& output : entry.second) {
            os << output << ", ";
        }
        os << "], ";
    }
    os << "}\n"
       << "tag2task: {";
    for (const auto& entry : task_info.tag2task) {
        os << entry.first << ": [";
        for (const auto& task : entry.second) {
            os << task << ", ";
        }
        os << "], ";
    }
    os << "}\n"
       << "task2tag: {";
    for (const auto& entry : task_info.task2tag) {
        os << entry.first << ": " << entry.second << ", ";
    }
    os << "}\n"
       << "has_input: ";
    for (const auto& task : task_info.has_input) {
        os << task << " ";
    }
    os << "\n"
       << "task2input: {";
    for (const auto& entry : task_info.task2input) {
        os << entry.first << ": [";
        for (const auto& input : entry.second) {
            os << input << ", ";
        }
        os << "], ";
    }
    os << "}\n"
       << "task2input_copy: {";
    for (const auto& entry : task_info.task2input_copy) {
        os << entry.first << ": [";
        for (const auto& input : entry.second) {
            os << input << ", ";
        }
        os << "], ";
    }
    os << "}\n"
       << "layer_topo_order: {";
    for (const auto& layer : task_info.layer_topo_order) {
        os << "[";
        for (const auto& task : layer) {
            os << task << ", ";
        }
        os << "], ";
    }
    os << "}\n"
       << "task2index: {";
    for (const auto& entry : task_info.task2index) {
        os << entry.first << ": (" << entry.second.first << ", " << entry.second.second << "), ";
    }
    os << "}\n";

    return os;
}

#if DEBUG
int main() {
    std::string pwd = ""; // Set your working directory
    std::string dag_file = pwd + "dag_dot_prof_file_3840_dmda.txt";
    TaskInfo task_info(dag_file);
    std::cout << task_info << std::endl;
    auto dep = task_info.get_dep();
    std::cout << "dep: " << std::endl;
    for (int t = 0; t < dep.size(); ++t) {
        std::cout << "timestamp: " << t << std::endl;
        for (int w = 0; w < dep[t].size(); ++w) {
            std::cout << "width: " << w << std::endl;
            for (const auto& task : dep[t][w]) {
                std::cout << "(" << task.first << ", " << task.second << "), ";
            }
            std::cout << std::endl;
        }
    }
    return 0;
}
#endif

File: starpu/parse_dag/parse_deps.py
import os
import copy


'''
    Parse dag file to get dependency
    Usage:
        task_info = TaskInfo(dag_file)
        task_info.get_dep()
'''
class TaskInfo:
    def __init__(self, dag_file):
        self.dag_file = dag_file
        self.all_task = []
        self.input_task = set() # task has no input
        self.output_task = set() # task has no output
        self.task2output = {} # task -> output
        self.tag2task = {} # tag -> task
        self.task2tag = {} # task -> tag
        self.has_input = set()
        self.task2input = {} # task -> input
        self.task2input_copy = {} # task -> input
        self.summary()

    def __str__(self):
        return 'dag_file: {}\nall_task: {}\ninput_task: {}\noutput_task: {}\ntask2output: {}\ntag2task: {}\ntask2tag: {}\nhas_input: {}'.format(
            self.dag_file, self.all_task, self.input_task, self.output_task, self.task2output, self.tag2task, self.task2tag, self.has_input
        )

    def _parse_dag(self):
        with open(self.dag_file, 'r') as f:
            lines = f.readlines()
            parse_tag = True
            split_line = -1
            for i, line in enumerate(lines):
                line = line.strip()
                if not line:
                    split_line = i
                    break
            tag_info = lines[:split_line]
            dep_info = lines[split_line + 1:]
            # tag line is like this
            # POTRF: [0, 10, 16, 19]
            # tag : [task1, task2, ...]
            for line in tag_info:
                line = line.strip()
                tag, tasks = line.split(':')
                tag = tag.strip()
                tasks = tasks.strip()
                tasks = tasks[1:-1].split(',')
                tasks = [task.strip() for task in tasks if task]
                self.tag2task[tag] = tasks
                for task in tasks:
                    self.task2tag[task] = tag
                self.all_task.extend(tasks)
                
            # dep line is like this
            # task1: (task2, task3, ...),
            # task : (output1, output2, ...),

            for line in dep_info:
                line = line.strip()
                if not line:
                    continue
                task, outputs = line.split(':')
                task = task.strip()
                outputs = outputs.strip()
                outputs = outputs[1:-2].split(',')
                outputs = [output.strip() for output in outputs if output]
                self.task2output[task] = outputs
                if not outputs:
                    self.output_task.add(task)
                for output in outputs:
                    self.has_input.add(output)
                    if output not in self.task2input:
                        self.task2input[output] = []
                    self.task2input[output].append(task)
            self.input_task = set(self.all_task) - self.has_input
            self.task2input_copy = copy.deepcopy(self.task2input)

    def _topo_sort(self):
        # topo sort
        # 1. find all task with no input
        # 2. remove the task from the graph
        # 3. repeat 1 and 2 until no task left
        # 4. Use bfs to get tasks of each level 

        # 1. find all task with no input
        # breakpoint()
        layer_topo_order = []
        topo_order = []
        queue = []
        for task in self.input_task:
            queue.append(task)
        while len(queue) > 0:
            n = len(queue)
            layer = []
            for i in range(n):
                task = queue.pop(0)
                layer.append(task)
                outputs = self.task2output[task]
                for output in outputs:
                    self.task2input[output].remove(task)
                    if not self.task2input[output]:
                        queue.append(output)
            layer_topo_order.append(layer)
        for layer in layer_topo_order:
            topo_order.extend(layer)
        return layer_topo_order, topo_order
    
    def _get_dep(self, task):
        if task in self.input_task:
            return [(-1, -1)]
        return [self.task2index[input] for input in self.task2input_copy[task]]

    def get_dep(self):
        # map layer_topo_order, element transform to self.task2index[self.task2input_copy[element]]
        return [[self._get_dep(task) for task in layer] for layer in self.layer_topo_order]

    def summary(self):
        self._parse_dag()
        self.layer_topo_order, _ = self._topo_sort()
        self._reindex_task()

    def get_timestamp(self):
        return len(self.layer_topo_order)
    
    def get_width_of_timestamp(self, t):
        return len(self.layer_topo_order[t])

    def get_task_of_timestamp(self, t):
        return self.layer_topo_order[t]
        
    def _reindex_task(self):
        # reindex from typo_layer_order
        # return task2index
        task2index = {}
        index2task = {}
        # index is a tuple (timestamp, width)
        for t, layer in enumerate(self.layer_topo_order):
            for w, task in enumerate(layer):
                task2index[task] = (t, w)
                index2task[(t, w)] = task
        self.task2index = task2index
    

if __name__ == '__main__':
    pwd = os.getcwd()
    dag_file = os.path.join(pwd, "dag_dot_prof_file_3840_dmda.txt")
    task_info = TaskInfo(dag_file)
    print(task_info)
    print(task_info.get_dep())
    # print(task_info)
    # layer_topo_order, topo_order = task_info.topo_sort()
    # print(layer_topo_order)
    # print(task_info.get_dep())